Fix user modification on Alpine

samdoran opened the pull request:
##### SUMMARY

Fixes #66679
Fixes #61911
Fixes #68676
Closes #84642

Fix user modification on Alpine. This change will handle primary group, shell, and password changes. It does not handle expiration changes, though it probably could with a bit more work.

The Alpine documentation recommends manually editing `/etc/password` to change the shell. These changes will rewrite the `/etc/passwd` file using `atomic_move` but creates a backup beforehand.

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

samdoran committed:
Update docs

samdoran committed:
Add changelog fragment

samdoran committed:
Modify user information on Alpine

Modify the passwd file directly when changing user shell. This is
the recommended approach by Alpine.

bcoca commented on lib/ansible/modules/user.py:
```diff
@@ -3244,6 +3264,42 @@ def modify_user(self):
+                with open(tmpfile, 'w') as f:
+                    f.writelines(contents)
+
+                self.module.atomic_move(tmpfile, self.PASSWORDFILE)
```
I would call `backup_local` right before this point.

samdoran replied:
I had that in there originally and was wondering if I should add that back in. `backup` isn't in the module parameters, so there's no way to turn it off. I'm not sure how folks would feel about that, but I suppose it's better safe than sorry.

samdoran replied:
`/etc/passwd` is backed up now.

samdoran committed:
Make account locking work correctly

samdoran committed:
Run more tests on Alpine

samdoran committed:
Create a backup /etc/passwd

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -3114,6 +3120,18 @@ class BusyBox(User):
+            password = current_password.lstrip('!')
+
+        # Ensure the account is locked at a minimum
+        result = '{lock}{password}'.format(lock=lock, password=password) or '!'
```
```suggestion
        result = f'{lock}{password}' or '!'
```

samdoran replied:
Yay! We can use format strings now.

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -3174,15 +3193,15 @@ def create_user(self):
             cmd = [self.module.get_bin_path('chpasswd', True)]
             cmd.append('--encrypted')
-            data = '{name}:{password}'.format(name=self.name, password=self.password)
+            data = '{name}:{password}'.format(name=self.name, password=self._build_password_string())
```
```suggestion
            data = f'{self.name}:{self._build_password_string()}'
```

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -3210,13 +3229,26 @@ def modify_user(self):
+        gid = user_info[3]
+        if self.group is not None:
+            if not self.group_exists(self.group):
+                self.module.fail_json(msg="Group %s does not exist" % self.group)
```
```suggestion
                self.module.fail_json(msg=f"Group {self.group} does not exist")
```

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -3235,14 +3267,57 @@ def modify_user(self):
+            ):
+                cmd = [self.module.get_bin_path('chpasswd', True)]
+                cmd.append('--encrypted')
+                data = '{name}:{password}'.format(name=self.name, password=new_password)
```
```suggestion
                data = f'{self.name}:{new_password}'
```

Akasurde commented:
@samdoran while you are here, can you incorporate https://github.com/ansible/ansible/pull/84642 changes as well?

samdoran committed:
Use fstrings.

samdoran committed:
Ensure password is a string.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3114,6 +3120,18 @@ class BusyBox(User):
         - modify_user()
     """
+    def _build_password_string(self, current_password=None):
+        lock = '!' if self.password_lock else ''
```
The `'!'` literal is used 3 times in this function. Perhaps, it's worth putting it into a named var?

webknjaz replied:
Oh, it's also in use outside, so maybe even a module-level constant.

samdoran replied:
An enum might be even better. I think adding a module-level constant would greatly increase the scope of this change, though. That could be done in another PR.

samdoran committed:
Use a variable for storing the lock indicator

samdoran committed:
Make conditional easier to read

samdoran committed:
Update changelogs

samdoran commented:
I made some changes to also fix #68786. It's tangentially related. BusyBox handles `!` at the beginning of the password hash a bit differently than other distros and will prevent logins if a `!` is present.

KalleDK commented on lib/ansible/modules/user.py:
```diff
@@ -3235,14 +3284,56 @@ def modify_user(self):
+            to_native(gid),
+            self.comment or user_info[4],
+            self.home or user_info[5],
+            self.shell or user_info[6],
```
Review from phone, so sorry if I missed it, but I can't see if you validate the shell against /etc/shells to validate that the shell exists?

samdoran replied:
That's correct.

samdoran replied:
Validating that the shell exists doesn't seem to be something that is done by other classes or the underlying tools themselves.

samdoran replied:
`adduser` will let you add any value for the shell. `chsh` only warns about a non-existent value but still returns 0.

KalleDK replied:
Then I see no problem with it, it would be nice to give a warning like chsh though

samdoran replied:
I can see about adding that but it may be better to handle that in a follow up PR.

KalleDK replied:
Again no problem :) Good idea with follow up pr

KalleDK replied:
Again no problem :) Good idea with follow up pr

samdoran replied:
https://github.com/ansible/ansible/pull/86342

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3114,6 +3120,36 @@ class BusyBox(User):
+        This method will return '*' at a minimum to avoid creating an enabled
+        account with no password.
+        """
+        lock_indicator = '!'
```
@samdoran why do you feel like making it a global constant would increase the scope of the PR? It'd be just this line moved elsewhere + used on an already edited line. Did you think you'd have to update lines you don't touch? I think this could be scoped to just the lines you update and adjusting the rest would be a follow-up.

samdoran replied:
That makes sense. I was thinking you meant to replace the value in all the classes in the module with the global constant, which would greatly increase the scope of the change.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3235,14 +3284,56 @@ def modify_user(self):
+        current_password = to_native(user_info[1])
+        new_password = self._build_password_string(current_password)
+        if self.update_password == 'always':
+            lock_status_mismatch = self.password_lock and not current_password.startswith('!')
```
With the earlier comment, I'd only expect this to be updated too:
```suggestion
            lock_status_mismatch = self.password_lock and not current_password.startswith(LOCK_INDICATOR)
```

samdoran replied:
I'll work on this tomorrow.

samdoran replied:
Changed!

samdoran committed:
Use global for lock indicator

Only used in the BusyBox class currently to keep the scope of these
changes focused.

samdoran committed:
Change the default password to a value that enables the account

The default behavior on Alpine is the create a locked account. In order
to enable the account, set the password to ‘*’ if no password was provided and
the account was not specified to be locked.

Shrews commented on lib/ansible/modules/user.py:
```diff
@@ -3246,14 +3295,55 @@ def modify_user(self):
+        change = False
+        with open(self.PASSWORDFILE, 'r') as password_file:
+            for line in password_file:
+                if line.startswith(self.name):
```
This seems like it would match multiple users. Perhaps needs to add ":" to the match string? Haven't looked at the existing tests, but it might be a good idea to add a test case to catch and validate that scenario.

samdoran replied:
Good catch!

samdoran replied:
I'm doing some testing but I think matching on `name:` is the correct solution.

samdoran replied:
Without that fix we get duplicate lines it `/etc/passwd`. Not good.

Shrews commented on lib/ansible/modules/user.py:
```diff
@@ -3246,14 +3295,55 @@ def modify_user(self):
+            rc = 0
+            if not self.module.check_mode:
+                tmpfd, tmpfile = tempfile.mkstemp(dir=self.module.tmpdir)
+                with open(tmpfile, 'w') as f:
```
Using `with os.fdopen()` instead of `with open()` will avoid the need for a second open file descriptor, and properly cleans up the open fd `tmpfd` for us, too.

samdoran replied:
Help me understand this a bit more.

Looking at the `fd.open()` docs it seems like it's an alias to `open` but accepts an int instead of a string.

The high level thing I'm trying to accomplish is opening a file, writing data to it, closing the file, then moving that file into place. Maybe I can do that without the second file descriptor but I'm not seeing how.

I can change it to this:
```python
with os.fdopen(tmpfd, 'w') as f:
...
```

But I don't see how that prevents a second open file descriptor. What am I missing?

Shrews replied:
`mkstemp()` returns an already opened fd which can be reused by `os.fdopen()`. Using `open()` with a string gives you a new fd. Probably best seen with a code example:

```python
tmpfd, tmpfile = tempfile.mkstemp()
print(f"{tmpfd=}, {tmpfile=}")

f = os.fdopen(tmpfd, "w")
print(f"{f.fileno()=}")    # same fileno as tmpfd

f2 = open(tmpfile, "w")
print(f"{f2.fileno()=}")   # new fileno, different from tmpfd
```

The gory details, if I followed the Cpython code correctly, the passed in fd gets reused [here](https://github.com/python/cpython/blob/d5cb9f6a9b6f48cc08c4422259498d4fd023357a/Lib/_pyio.py#L1537). Otherwise, if we pass a string, a new one get created [here](https://github.com/python/cpython/blob/d5cb9f6a9b6f48cc08c4422259498d4fd023357a/Lib/_pyio.py#L1590).

What you have works, as everything eventually should get cleaned up in the end when everything terminates (that same pattern is used in soooo many places in the code base already). Was just offering a more OCD initiated must-clean-up-my-own-mess solution.  😄

samdoran replied:
Ok, I did some more reading and looked at this with fresh eyes this morning. I see how `os.fdopen()` avoids opening a second file descriptor now plus it will actually close it avoid the leak this code currently has.

samdoran replied:
I wasn't familiar with `os.fdopen`. Thanks for sharing that and thank you for the great explanation. It's always a good day when you have to dip into CPython code. 😃

samdoran committed:
Match name plus delimiter to avoid incorrect changes

samdoran committed:
Use os.fdopen to properly close the temp file file descriptor

Without this change a secord file descriptor is used and the tmpfd descriptor
is never closed, resulting in a file descriptor leak.

webknjaz commented:
Got a nightly failure under Alpine 3.23 in the `become_unprivileged` test: https://dev.azure.com/ansible/ansible/_build/results?buildId=170700&view=logs&j=1721704d-3ce2-5d7c-7134-6be08f87ee74&t=33a78053-86e3-5aea-37eb-8744f958ec00&l=4811. Re-running to see if it's a fluke.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3221,13 +3257,26 @@ def modify_user(self):
+
+        gid = user_info[3]
+        if self.group is not None:
+            if not self.group_exists(self.group):
+                self.module.fail_json(msg=f'Group {self.group} does not exist')
+
+            group_info = self.group_info(self.group)
+            if group_info:
+                gid = group_info[2]
```
Looks like this isn't covered in tests.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3221,13 +3257,26 @@ def modify_user(self):
+        user_info = self.user_info()
+
+        if not user_info:
+            return rc, out, err
```
This isn't covered either.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3246,14 +3295,55 @@ def modify_user(self):
+                if line.startswith('%s:' % self.name):
+                    fields = line.strip().split(':')
+                    if fields != passwd_entry:
+                        change = True
+                        line = ':'.join(passwd_entry) + '\n'
```
This never happens in tests.

webknjaz commented on lib/ansible/modules/user.py:
```diff
@@ -3246,14 +3295,55 @@ def modify_user(self):
+                contents.append(line)
+
+        if change:
+            rc = 0
+            if not self.module.check_mode:
+                tmpfd, tmpfile = tempfile.mkstemp(dir=self.module.tmpdir)
+                with os.fdopen(tmpfd, 'w') as f:
+                    f.writelines(contents)
+
+                self.module.backup_local(self.PASSWORDFILE)
+                self.module.atomic_move(tmpfile, self.PASSWORDFILE)
```
And this isn't covered either.
